tutorials-v5/time-evolution/013_nonmarkovian_monte_carlo.md

```diff
@@ -0,0 +1,468 @@
+a1 = x1 / ((x1 - x2) * (x1 - x3))
+a2 = x2 / ((x2 - x1) * (x2 - x3))
+a3 = x3 / ((x3 - x2) * (x3 - x1))
+y1 = (x1**2) ** (1 / 2)
+y2 = (x2**2) ** (1 / 2)
+y3 = (x3**2) ** (1 / 2)
+
+
+def c(t):
+    return (
+        2 * x1 * a1 * np.exp((beta * x1**2 + 1j * delta) * t)
+        + a2 * (x2 + y2) * np.exp((beta * x2**2 + 1j * delta) * t)
+        - (a1 * y1 * (1 - special.erf((beta * x1**2 * t) ** (1 / 2)))
+           * np.exp((beta * x1**2 + 1j * delta) * t))
+        - (a2 * y2 * (1 - special.erf((beta * x2**2 * t) ** (1 / 2)))
+           * np.exp((beta * x2**2 + 1j * delta) * t))
+        - (a3 * y3 * (1 - special.erf((beta * x3**2 * t) ** (1 / 2)))
+           * np.exp((beta * x3**2 + 1j * delta) * t))
+    )
+
+
+def cd(t):  # time derivative
+    return (
+        ((beta * x1**2 + 1j * delta) * 2 * x1 * a1
+         * np.exp((beta * x1**2 + 1j * delta) * t))
+        + ((beta * x2**2 + 1j * delta) * a2 * (x2 + y2)
+           * np.exp((beta * x2**2 + 1j * delta) * t))
+        - ((beta * x1**2 + 1j * delta) * a1 * y1
+           * (1 - special.erf((beta * x1**2 * t) ** (1 / 2)))
+           * np.exp((beta * x1**2 + 1j * delta) * t))
+        - ((beta * x2**2 + 1j * delta) * a2 * y2
+           * (1 - special.erf((beta * x2**2 * t) ** (1 / 2)))
+           * np.exp((beta * x2**2 + 1j * delta) * t))
+        - ((beta * x3**2 + 1j * delta) * a3 * y3
+           * (1 - special.erf((beta * x3**2 * t) ** (1 / 2)))
+           * np.exp((beta * x3**2 + 1j * delta) * t))
+        + (a1 * y1 * x1 * np.exp(-beta * t * x1**2)
+           * (beta / (t * np.pi + 0.00001)) ** (1 / 2)
+           * np.exp((beta * x1**2 + 1j * delta) * t))
+        + (a2 * y2 * x2 * np.exp(-beta * t * x2**2)
+           * (beta / (t * np.pi + 0.00001)) ** (1 / 2)
+           * np.exp((beta * x2**2 + 1j * delta) * t))
+        + (a3 * y3 * x3 * np.exp(-beta * t * x3**2)
+           * (beta / (t * np.pi + 0.00001)) ** (1 / 2)
+           * np.exp((beta * x3**2 + 1j * delta) * t))
+    )
+
+
+def S(t):
+    return -2 * np.imag(cd(t) / c(t))
+
+
+def Gamma(t):
+    return -2 * np.real(cd(t) / c(t))
+```
+
+Define the time interval. The initial time is shifted to avoid negative values of $\Gamma(t)$ at $t=0$ (which is unphysical):
+
+```python
+ti = optimize.root_scalar(Gamma, bracket=(1.4, 1.5)).root
+duration = 10
+steps = 100
+
+times = np.linspace(ti, ti + duration, steps + 1)
+```
+
+We plot the functions $\Gamma(t)$ and $S(t)$ over the time interval, demonstrating that $\Gamma(t)$ becomes negative:
+
+```python
+Gamma_values = np.array([Gamma(t) for t in times])
+S_values = np.array([S(t) for t in times])
+
+plt.plot(times - ti, Gamma_values, label=r"$\Gamma(t)$")
+plt.plot(times - ti, S_values, label=r"$S(t)$")
+plt.xlabel(r"$t$")
+plt.legend()
+plt.show()
+```
+
+To speed up the following calculation by a factor 3-4, we store $\Gamma(t)$ and $S(t)$ as interpolations.
+
+```python
+Gamma_tmp = CubicSpline(times, Gamma_values)
+S_tmp = CubicSpline(times, S_values)
```

Comment:
Done. (I don't call `qutip.coefficient` explicitly because that is done by `nm_mcsolve`.)
Yeah, `qutip.coefficient` not liking real-valued splines was the reason why I had wrapped the splines in functions in the previous version.